olp-cpp-sdk-dataservice-read/tests/StreamLayerClientImplTest.cpp

```diff
@@ -0,0 +1,303 @@
+    ASSERT_FALSE(response.IsSuccessful());
+
+    Mock::VerifyAndClearExpectations(network_mock.get());
+  }
+}
+
+TEST(StreamLayerClientImplTest, SubcribeCancellableFuture) {
+  auto network_mock = std::make_shared<NetworkMock>();
+  auto cache_mock = std::make_shared<CacheMock>();
+  olp::client::OlpClientSettings settings;
+  settings.network_request_handler = network_mock;
+  settings.cache = cache_mock;
+
+  {
+    SCOPED_TRACE("Subscribe success");
+
+    StreamLayerClientImpl client(kHRN, kLayerId, settings);
+
+    SetupNetworkExpectation(*network_mock, kUrlLookupStream,
+                            kHttpResponseLookupStream,
+                            olp::http::HttpStatusCode::OK);
+
+    SetupNetworkExpectation(
+        *network_mock, kUrlStreamSubscribe, kHttpResponseSubscribe,
+        olp::http::HttpStatusCode::CREATED, RequestMethod::POST);
+
+    auto future = client.Subscribe(SubscribeRequest()).GetFuture();
+
+    EXPECT_EQ(future.wait_for(kTimeout), std::future_status::ready);
+
+    const auto& response = future.get();
+    ASSERT_TRUE(response.IsSuccessful());
+
+    ASSERT_EQ(response.GetResult(), kSubcriptionId);
+
+    Mock::VerifyAndClearExpectations(network_mock.get());
+  }
+
+  {
+    SCOPED_TRACE("The second subscribe");
+
+    StreamLayerClientImpl client(kHRN, kLayerId, settings);
+
+    SetupNetworkExpectation(*network_mock, kUrlLookupStream,
+                            kHttpResponseLookupStream,
+                            olp::http::HttpStatusCode::OK);
+
+    SetupNetworkExpectation(
+        *network_mock, kUrlStreamSubscribe, kHttpResponseSubscribe,
+        olp::http::HttpStatusCode::CREATED, RequestMethod::POST);
+
+    {
+      auto future = client.Subscribe(SubscribeRequest()).GetFuture();
+
+      EXPECT_EQ(future.wait_for(kTimeout), std::future_status::ready);
+
+      const auto& response = future.get();
+      ASSERT_TRUE(response.IsSuccessful());
+
+      ASSERT_EQ(response.GetResult(), kSubcriptionId);
+    }
+    {
+      auto future = client.Subscribe(SubscribeRequest()).GetFuture();
+
+      EXPECT_EQ(future.wait_for(kTimeout), std::future_status::ready);
+
+      const auto& response = future.get();
+      ASSERT_FALSE(response.IsSuccessful());
+
+      ASSERT_EQ(response.GetError().GetErrorCode(),
+                olp::client::ErrorCode::InvalidArgument);
+    }
+    Mock::VerifyAndClearExpectations(network_mock.get());
+  }
+}
+
+TEST(StreamLayerClientImplTest, SubrcribeCancel) {
+  auto network_mock = std::make_shared<NetworkMock>();
+  auto cache_mock = std::make_shared<CacheMock>();
+  olp::client::OlpClientSettings settings;
+  settings.network_request_handler = network_mock;
+  settings.cache = cache_mock;
+  settings.task_scheduler =
+      olp::client::OlpClientSettingsFactory::CreateDefaultTaskScheduler(1);
+
+  {
+    // Simulate a loaded queue
+    settings.task_scheduler->ScheduleTask(
+        []() { std::this_thread::sleep_for(std::chrono::seconds(1)); });
```

Comment:
I think it is OK (however there are some doubts about scheduling of the main thread's work), but I'm still encourage you to use `GenerateNetworkMockActions` from `NetworkMock.h` file, because with that way, it is easier to control and predict threads execution.

Reply:
Replaced with promise+future approach, it should be safer. Though, left SubrcribeCancelOnClientDestroy with sleep since not possible to set a value to the promise before a desctructor call.